test/helper/e2e2/kube/kube.go

```diff
@@ -0,0 +1,178 @@
+	key := client.ObjectKeyFromObject(obj)
+	old := obj.DeepCopyObject().(client.Object)
+	err := kubeClient.Get(ctx, key, old)
+	switch {
+	case err == nil:
+		obj = obj.DeepCopyObject().(client.Object)
+		obj.SetResourceVersion(old.GetResourceVersion())
+		if err := kubeClient.Update(ctx, obj); err != nil {
+			return fmt.Errorf("failed to update %s: %w", key, err)
+		}
+	case apierrors.IsNotFound(err):
+		if err := kubeClient.Create(ctx, obj); err != nil {
+			return fmt.Errorf("failed to create %s: %w", key, err)
+		}
+	default:
+		return fmt.Errorf("failed to apply %s: %w", key, err)
+	}
+	return nil
+}
+
+func CreateNamespace(ctx context.Context, kubeClient client.Client, namespace string) error {
+	return kubeClient.Create(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}})
+}
+
+func WipeNamespace(ctx context.Context, kubeClient client.Client, namespace string) error {
+	return kubeClient.Delete(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}})
+}
+
+func HasNamespace(ctx context.Context, kubeClient client.Client, namespace string) bool {
+	return kubeClient.Get(ctx, client.ObjectKey{Name: namespace}, &corev1.Namespace{}) == nil
+}
+
+func AssertObjReady(ctx context.Context, kubeClient client.Client, key client.ObjectKey, obj ObjectWithStatus) (bool, error) {
+	err := kubeClient.Get(ctx, key, obj)
+	if err != nil {
+		return false, fmt.Errorf("failed to get object %v: %w", key, err)
+	}
+	for _, condition := range obj.GetConditions() {
+		if condition.Type == "Ready" && condition.Status == metav1.ConditionTrue {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func AssertObjExists(ctx context.Context, kubeClient client.Client, key client.ObjectKey, obj client.Object) (bool, error) {
+	err := kubeClient.Get(ctx, key, obj)
+	if err != nil {
+		return false, fmt.Errorf("failed to get object %v: %w", key, err)
+	}
+	return true, nil
+}
+
+func SetNamespace(obj client.Object, ns string) client.Object {
+	renamed := obj.DeepCopyObject().(client.Object)
+	renamed.SetNamespace(ns)
+	return renamed
+}
+
+func getTestScheme(addToSchemeFunctions ...func(*runtime.Scheme) error) (*runtime.Scheme, error) {
+	testScheme := runtime.NewScheme()
+	for _, addToSchemeFn := range addToSchemeFunctions {
+		if err := addToSchemeFn(testScheme); err != nil {
+			return nil, fmt.Errorf("failed to add to testScheme: %w", err)
+		}
+	}
+	return testScheme, nil
+}
+
+func getKubeClient(scheme *runtime.Scheme) (client.Client, error) {
+	restCfg, err := ctrl.GetConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Kubernetes config (is cluster configured?): %w", err)
+	}
+	kubeClient, err := client.New(restCfg, client.Options{Scheme: scheme})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Kubernetes client (is cluster up?): %w", err)
+	}
+	return kubeClient, nil
+}
+
+func assertCRD(ctx context.Context, kubeClient client.Client, targetCRD string) error {
```

Comment:
what is `targetCRD`? is it a GVK, a GVR, something else? I suggest passing the concrete type.

Reply:
will use the type instead of the name